distributor/MarketDistributor.ts

```diff
@@ -0,0 +1,161 @@
+import { EtherscanBlockFetcher } from "./blockFetcher/EtherscanBlockFetcher";
+
+export interface IMarketStorage {
+  storeMarketsEmissions: (ageEpoch: AgeEpoch, params: Awaited<ReturnType<DistributionFn>>) => Promise<void>;
+}
+export interface IBlockFetcher {
+  blockFromTimestamp: (timestamp: number, direction: "before" | "after") => Promise<number>;
+}
+
+export interface AgeEpoch {
+  id: string;
+  initialTimestamp: string;
+  finalTimestamp: string;
+  snapshotBlock?: number;
+  distributionScript: string;
+  distributionParameters: DistributionParameters;
+}
+interface DistributionParameters extends Record<string, any> {
+  totalEmission: string;
+}
+
+export interface IConfigurationFetcher {
+  getConfigurations: (ids: string[]) => AgeEpoch[];
+  getAllConfigurations: () => AgeEpoch[];
+  getEndedConfigurations: (ts: number) => AgeEpoch[];
+  getAccumulatedRewardsEmitted: (id: string) => BigNumber;
+}
+export class InMemoryConfigurationFetcher implements IConfigurationFetcher {
+  getConfigurations(ids: string[]) {
+    return ageEpochs.filter(({ id }) => ids.includes(id));
+  }
+  getAllConfigurations() {
+    return ageEpochs;
+  }
+  static getEndedConfigurations(ts: number) {
+    return ageEpochs.filter(({ finalTimestamp }) => ts > parseDate(finalTimestamp));
+  }
+  getEndedConfigurations(ts: number) {
+    return InMemoryConfigurationFetcher.getEndedConfigurations(ts);
+  }
+
+  getAccumulatedRewardsEmitted(id: string) {
+    const lastEpoch = ageEpochs.find(({ id: _id }) => _id === id);
+    if (!lastEpoch) throw new Error(`No epoch found for id ${id}`);
+    const allEmitted = ageEpochs.filter(
+      ({ initialTimestamp }) => parseDate(initialTimestamp) < parseDate(lastEpoch.finalTimestamp)
+    );
+
+    return allEmitted.reduce((acc, { distributionParameters: { totalEmission } }) => {
+      return acc.add(parseUnits(totalEmission));
+    }, constants.Zero);
+  }
+}
+
+export class MarketDistributor {
+  #provider: providers.Provider;
+  #blockFetcher: IBlockFetcher;
+  #configFetcher: IConfigurationFetcher;
+
+  constructor(
+    _provider: providers.Provider,
+    _blockFetcher: IBlockFetcher = new EtherscanBlockFetcher(process.env.ETHERSCAN_API_KEY!),
+    _configFetcher: IConfigurationFetcher = new InMemoryConfigurationFetcher()
+  ) {
+    this.#provider = _provider;
+    this.#blockFetcher = _blockFetcher;
+    this.#configFetcher = _configFetcher;
+  }
+
+  public async distribute(ids: string[] = []) {
+    const configurations =
+      ids.length > 0 ? this.#configFetcher.getConfigurations(ids) : this.#configFetcher.getAllConfigurations();
+    const results: Record<string, Awaited<ReturnType<DistributionFn>>> = {};
+    for (const ageEpoch of configurations) {
+      if (!isCorrectConfig(ageEpoch)) throw new Error(`Invalid age epoch config: ${ageEpoch}`);
+      console.log("Distributing", ageEpoch.id);
+      const distributionFn = (await import(`./distributionScripts/${ageEpoch.distributionScript}.ts`).then(
```

Comment:
Importing a TS file? I file like it's not what we want, unless we're running everything in TS Node.